Define a new tool for generating alert trees, and use it for MetalK8s

gdemonet opened the pull request:
**Component**: monitoring, build, salt

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: In the Overview page, we will require higher-level logical alerts to display the health of each "service category".

**Summary**:

- [x] Build a Python library for creating alert hierarchies and manipulating them
- [x] Use this library to represent all required alerts for MetalK8s
- [x] Generate the rules at build-time
- [x] Deploy the generated rules
- [x] Document the library in the developer guide

**Acceptance criteria**: 

- Overview page works after a plain bootstrap
- Other projects can use this library to build their own tree

---

#### `tox -e alert-tree -- show`

![image](https://user-images.githubusercontent.com/45066113/133753267-92f69987-dee6-4409-8703-49d61ef75b30.png)

#### `tox -e alert-tree -- gen-rule`

![image](https://user-images.githubusercontent.com/45066113/133766038-4dfc8699-6a60-4f8f-9e7b-8c9bbfa15484.png)

---

Heavily inspired by an internal tool developed by @sayf-eddine-scality

gdemonet committed:
tools: Initialize a `lib_alert_tree` Poetry package

This will serve as a shareable library for generating and manipulating
hierarchies of Prometheus alerts in various contexts.

gdemonet committed:
tools/alert-tree: Add serializable data containers

This will facilitate the rendering of objects into valid YAML manifests,
but shouldn't be directly created by users unless strictly required.

gdemonet committed:
tools/alert-tree: Define models for building trees

Using `ExistingAlert` and `DerivedAlert`, one can build a hierarchy of
alerts, with automatic definition of the PromQL expression required to
derive an alert from its children, and some other helpers for
manipulating such alert hierarchies.

Each one maps to an `AlertRule` serializable container, although
`ExistingAlert`s are not intended to be used like this, but only as a
simplified container of an alert name and its labels. This allows to
build complete PromQL queries when referencing such alerts in their
parent definitions.

gdemonet committed:
tools/alert-tree: Add Kubernetes helpers

Some Kubernetes objects have a number of related alerts, which can be
useful to select when describing a higher-level entity (e.g. combining
application-level alerts with these ones can give an overview of the
service health).

gdemonet committed:
tools/alert-tree: Generate CLI from root alerts

This allows easy exploration of alert trees and rendering of
PrometheusRule manifests.

gdemonet committed:
alert-tree: Define a full alert tree for MetalK8s

This hierarchy is built using the `tools/lib-alert-tree` library, and
can be easily inspected using `tox -e alert-tree show` (requires Poetry
to be installed on the system).

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @lucieleonard

* @teddyandrieux

* @thomasdanan

* @nicolast

* @chengyanjin

* @jbwatenbergscality

* @gdemonet

* @alexandre-allard-scality

gdemonet committed:
build: Check 'tox' external command

This will provide cleaner errors in case it's missing.

gdemonet committed:
build: Clean some task names and documentation

Some task names were misleading as to what the task was doing (mostly
for tasks generating files).

gdemonet committed:
build,salt: Render and deploy the full alert tree

The definition of all this tree lives under
`tools/lib-alert-tree/metalk8s`.

gdemonet committed:
doc/developer: Update info about buildchain logs

gdemonet committed:
docs/developer: Document `tools/lib-alert-tree`

Also include a tox environment to export the library dependencies in the
requirements.txt format, which gets included by docs/requirements.in, so
that we can rely on `autodoc` to generate a reference documentation of
the library.

gdemonet commented on buildchain/buildchain/config.py:
```diff
@@ -63,6 +63,7 @@ class ExtCommand(enum.Enum):
     VAGRANT = os.getenv("VAGRANT_BIN", "vagrant")
     OPERATOR_SDK = os.getenv("OPERATOR_SDK_BIN", "operator-sdk")
     GOFMT = os.getenv("GOFMT_BIN", "gofmt")
+    TOX = os.getenv("TOX_BIN", "tox")
```
If you wonder why I added this check, I started by using `tox` for rendering the rules manifest, but it was strange to add a requirement for both `tox` and `poetry` to be available for building the product, just for this small manifest. So I went with Docker instead :roll_eyes:

gdemonet commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @lucieleonard

* @teddyandrieux

* @thomasdanan

* @nicolast

* @chengyanjin

* @jbwatenbergscality

* @gdemonet

* @alexandre-allard-scality










*The following options are set:* **approve**

gdemonet committed:
build,docs: Remove PDF rendering with LaTeX

Rendering PDFs was already cumbersome, and broke with the changes
introduced by #3540.
It was decided that HTML would be the only format supported from now
onwards. The full HTML output is now part of the ISO, making it easy to
read offline (e.g. using `python -m http.server`).

sayf-eddine-scality commented on docs/developer/tools/lib-alert-tree/examples/example.py:
```diff
@@ -0,0 +1,44 @@
+BAR_BAZ_WARNING = Existing.warning("BarAlmostOutOfSpace", bar="baz")
+BAR_BAZ_CRITICAL = Existing.critical("BarAlmostOutOfSpace", bar="baz")
+
+# Derived alerts are built from a list of children and a relationship type (the default
```
imo it is gonna be hard to undertand the code with such implicit relationship

gdemonet replied:
How is it implicit? I feel it to be simpler to understand when you link all children from a parent, rather than linking the parent to all children individually. What would you change?

Or do you mean the default being set to "ANY"?

sayf-eddine-scality replied:
yes the ANY part

gdemonet replied:
OK I'll set it back to being a mandatory argument.

gdemonet replied:
Done in 542b283.

sayf-eddine-scality commented on docs/developer/tools/lib-alert-tree/index.rst:
```diff
@@ -0,0 +1,53 @@
+``lib_alert_tree`` Python library
+=================================
+
+This Python library serves as common grounds when composing Prometheus alerts
```
```suggestion
This Python library serves as a common ground when composing Prometheus alerts
```

gdemonet replied:
Done

sayf-eddine-scality commented on docs/developer/tools/lib-alert-tree/install.rst:
```diff
@@ -0,0 +1,69 @@
+.. code:: console
+
+   $ python3 -m pip install --use-pep517 \
+       "git+https://github.com/scality/metalk8s.git@feature/lib-alert-tree#egg=lib_alert_tree&subdirectory=tools/lib-alert-tree"
```
you will change the branch later? :D

gdemonet replied:
Done (removed the branch specification entirely, this way the command defaults to the default branch, I think... anyway, not really recommended to use this unless you know what you're doing)

alexandre-allard commented on docs/developer/tools/lib-alert-tree/usage.rst:
```diff
@@ -0,0 +1,52 @@
+``DerivedAlert`` instances are part of this manifest (exiting alerts should not
+need any additional rules, and are only referred to in the derived ``expr``).
+
+Use the ``-h`` / ``--help`` option to print a help message and discover more
```
There is no such `-h` option, only `--help`.

gdemonet replied:
Fixed

alexandre-allard reviewed (commented):
LGTM, nothing much to say, I've tested it and it's really cool.
I'm just wondering, you've changed some parent alert names (e.g. K8sMasterServicesDegraded > KubernetesControlPlaneDegraded), I don't know how these alerts are consumed UI-wise, etc. is it OK ? Or I assume you'll do a following PR on ARTESCA side to adapt that (at leas to remove the Prometheus rules deployment).
I don't approve yet since you already approved and you may want to edit some stuff (regarding the comments) prior to merge it.

gdemonet commented:
> LGTM, nothing much to say, I've tested it and it's really cool.

:heart:

> I'm just wondering, you've changed some parent alert names (e.g. K8sMasterServicesDegraded > KubernetesControlPlaneDegraded), I don't know how these alerts are consumed UI-wise, etc. is it OK ?

Good catch, I forgot about that, indeed I'll likely need to change a few things in MetalK8s UI.

> Or I assume you'll do a following PR on ARTESCA side to adapt that (at leas to remove the Prometheus rules deployment).

Indeed, one to remove the previous MetalK8s rules and maybe add some for ARTESCA services on top of those.

> I don't approve yet since you already approved and you may want to edit some stuff (regarding the comments) prior to merge it.

Good catch as well, I removed my `/approve`.

gdemonet committed:
shell-ui: Update alert names from new alert tree

Some names were changed while developing lib-alert-tree and the MetalK8s
alerts.

gdemonet committed:
tools/alert-tree: Make 'relationship' mandatory

This will keep the code more explicit.

gdemonet committed:
ci: Run lib-alert-tree unit-tests in pre-merge

gdemonet commented:
@alexandre-allard-scality I updated Shell UI's library for showing alerts in 89f5d12 (cc @JBWatenbergScality if you want to check I didn't miss anything).

gdemonet commented:
/approve

JBWatenbergScality commented on shell-ui/src/alerts/alertHooks.js:
```diff
@@ -6,7 +6,7 @@ import type { FilterLabels, Alert } from './services/alertUtils';
 
 export const getPlatformAlertSelectors = (): FilterLabels => {
-  return { alertname: ['PlatformAtRisk', 'PlatformDegraded'] };
+  return { alertname: ['ClusterAtRisk', 'ClusterDegraded'] };
```
I'm wondering wether we should not include both the old alert names and the new ones in this filter. Because when upgrading we could loose historical alerts in the global health component. But maybe this is a none existing issue if no one is using those alerts in prod yet.

gdemonet replied:
This lib hasn't been released yet, right? So I'm not sure how upgrade is a concern.
Some deployments may have alert rules matching the old names in place, and we may not remove them, but the lib won't care and won't fetch them.

JBWatenbergScality replied:
The lib is not yet released indeed, however we already have packaged those alerts in products that are released and my concern is that for such installations an upgrade to Metalk8s 2.11 bundling the `dashboard page` would reflect a wrong history of alerts (pre upgrade some `PlatformAtRisk` or `PlatformDegraded` could be raised and then post upgrade we will watch `ClusterAtRisk` and `ClusterDegraded`) which will then mislead the user experience

gdemonet replied:
Mmh, I consider it a feature addition, the previous alerts were not part of MetalK8s offering, so I don't think we should care.

JBWatenbergScality commented:
I think an update of `alert-grouping.rst` documentation is missing with the new alert names

JBWatenbergScality reviewed (approved):
Documentation will be done later on. Approving for now as discussed.

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.11`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.11`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye gdemonet.
